translib/api_tests_app.go

```diff
@@ -0,0 +1,188 @@
+////////////////////////////////////////////////////////////////////////////////
+//                                                                            //
+//  Copyright 2020 Broadcom. The term Broadcom refers to Broadcom Inc. and/or //
+//  its subsidiaries.                                                         //
+//                                                                            //
+//  Licensed under the Apache License, Version 2.0 (the "License");           //
+//  you may not use this file except in compliance with the License.          //
+//  You may obtain a copy of the License at                                   //
+//                                                                            //
+//     http://www.apache.org/licenses/LICENSE-2.0                             //
+//                                                                            //
+//  Unless required by applicable law or agreed to in writing, software       //
+//  distributed under the License is distributed on an "AS IS" BASIS,         //
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  //
+//  See the License for the specific language governing permissions and       //
+//  limitations under the License.                                            //
+//                                                                            //
+////////////////////////////////////////////////////////////////////////////////
+
+// +build test
+
+package translib
+
+import (
+	"encoding/json"
+	"reflect"
+	"strings"
+
+	"github.com/Azure/sonic-mgmt-common/translib/db"
+	"github.com/Azure/sonic-mgmt-common/translib/tlerr"
+	"github.com/golang/glog"
+)
+
+// apiTests is an app module for testing translib APIs.
+// Implements dummy handlers for paths starting with "/api-tests:".
+// Returns error if path contains "/error/"; see getError function.
+type apiTests struct {
+	path string
+	body []byte
+	appOptions
+
+	echoMsg string
+	echoErr string
+}
+
+func init() {
+	err := register("/api-tests:",
+		&appInfo{
+			appType:       reflect.TypeOf(apiTests{}),
+			isNative:      true,
+			tablesToWatch: nil})
+
+	if err != nil {
+		glog.Fatalf("Failed to register ApiTest app; %v", err)
+	}
+}
+
+func (app *apiTests) initialize(inp appData) {
```

Comment:
Where is the code that invokes this class/struct ?
Basically, where/who is the consumer of this struct ?

Reply:
This is called by translib infra when any request is made for path starting with "/api-tests:". Test code can use this to simulate success and various errors from app module. Details are in PR description.
I am currently using it in REST test code -- https://github.com/Azure/sonic-mgmt-framework/pull/52.